From 2390976439c31cbb8dfd6434c1461f25ae311fac Mon Sep 17 00:00:00 2001
From: Zhi Chen <quic_zhichen@quicinc.com>
Date: Tue, 26 Dec 2023 10:53:59 +0800
Subject: [PATCH] ath9k: fixed wrong queues in MCC mode

In MCC mode, wrong iee80211_queue were scheduled and frames were sent
in wrong channel.

Signed-off-by: Zhi Chen <quic_zhichen@quicinc.com>
---
 drivers/net/wireless/ath/ath9k/init.c |  6 +++++-
 drivers/net/wireless/ath/ath9k/xmit.c | 12 ++++++++++++
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath9k/init.c b/drivers/net/wireless/ath/ath9k/init.c
index 5388a0c..fd37765 100644
--- a/drivers/net/wireless/ath/ath9k/init.c
+++ b/drivers/net/wireless/ath/ath9k/init.c
@@ -1051,7 +1051,11 @@ static void ath9k_set_hw_capab(struct ath_softc *sc, struct ieee80211_hw *hw)
 	SET_IEEE80211_PERM_ADDR(hw, common->macaddr);
 
 	wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_CQM_RSSI_LIST);
-	wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_AIRTIME_FAIRNESS);
+
+	if (!ath9k_is_chanctx_enabled()) {
+		wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_AIRTIME_FAIRNESS);
+	}
+
 	wiphy_ext_feature_set(hw->wiphy,
 			      NL80211_EXT_FEATURE_MULTICAST_REGISTRATIONS);
 	wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_CAN_REPLACE_PTK0);
diff --git a/drivers/net/wireless/ath/ath9k/xmit.c b/drivers/net/wireless/ath/ath9k/xmit.c
index d6e5abc..221e221 100644
--- a/drivers/net/wireless/ath/ath9k/xmit.c
+++ b/drivers/net/wireless/ath/ath9k/xmit.c
@@ -1942,6 +1942,18 @@ void ath_txq_schedule(struct ath_softc *sc, struct ath_txq *txq)
 	while ((queue = ieee80211_next_txq(hw, txq->mac80211_qnum))) {
 		bool force;
 
+		if (ath9k_is_chanctx_enabled()) {
+			struct ieee80211_vif *vif;
+			struct ath_vif * avp;
+
+			vif = queue->vif;
+			avp = (struct ath_vif *)vif->drv_priv;
+			if (avp->chanctx != sc->cur_chan) {
+				ieee80211_return_txq(hw, queue, true);
+				continue;
+			}
+		}
+
 		tid = (struct ath_atx_tid *)queue->drv_priv;
 
 		ret = ath_tx_sched_aggr(sc, txq, tid);
-- 
2.7.4

